bots/task/__init__.py

```diff
@@ -0,0 +1,354 @@
+# our GitHub interacition.
+
+import argparse
+import os
+import random
+import socket
+import subprocess
+import sys
+import time
+import traceback
+
+sys.dont_write_bytecode = True
+
+import github
+import sink
+
+__all__ = (
+    "main",
+    "run",
+    "pull",
+    "issue",
+    "verbose",
+    "stale",
+)
+
+api = github.GitHub()
+verbose = False
+
+BOTS = os.path.normpath(os.path.join(os.path.dirname(__file__), ".."))
+BASE = os.path.normpath(os.path.join(BOTS, ".."))
+
+#
+# The main function takes a list of tasks, each of wihch has the following
+# fields, some of which have defaults:
+#
+#   title: The title for the task
+#   function: The function to call for the task
+#   options=[]: A list of string options to pass to the function argument
+#
+# The function for the task will be called with all the context for the task.
+# In addition it will be called with named arguments for all other task fields
+# and additional fields such as |verbose|. It should return a zero or None value
+# if successful, and a string or non-zero value if unsuccessful.
+#
+#   def run(context, verbose=False, **kwargs):
+#       if verbose:
+#           sys.stderr.write(image + "\n")
+#       return 0
+#
+# Call the task.main() as the entry point of your script in one of these ways:
+#
+#   # As a single task
+#   task.main(title="My title", function=run)
+#
+
+def main(**kwargs):
+    global verbose
+
+    task = kwargs.copy()
+
+    # Figure out a descriptoin for the --help
+    task["name"] = named(task)
+
+    parser = argparse.ArgumentParser(description=task.get("title", task["name"]))
+    parser.add_argument("-v", "--verbose", action="store_true", help="Verbose output")
+    parser.add_argument("--issue", dest="issue", action="store",
+        help="Act on an already created task issue")
+    parser.add_argument("--publish", dest="publish", default=os.environ.get("TEST_PUBLISH", ""),
+        action="store", help="Publish results centrally to a sink")
+    parser.add_argument("context", nargs="?")
+
+    opts = parser.parse_args()
+    verbose = opts.verbose
+
+    ret = 0
+
+    if "verbose" not in task:
+        task["verbose"] = opts.verbose
+    task["issue"] = opts.issue
+    task["publish"] = opts.publish
+
+    ret = run(opts.context, **task)
+
+    if ret:
+        sys.stderr.write("{0}: {1}\n".format(task["name"], ret))
+
+    sys.exit(ret and 1 or 0)
+
+def named(task):
+    if "name" in task:
+        return task["name"]
+    else:
+        return os.path.basename(os.path.realpath(sys.argv[0]))
+
+def begin(publish, name, context, issue):
+    if not publish:
+        return None
+
+    hostname = socket.gethostname().split(".")[0]
+    current = time.strftime('%Y%m%d-%H%M%M')
+
+    # Update the body for an existing issue
+    if issue:
+        number = issue["number"]
+        identifier = "{0}-{1}-{2}".format(name, number, current)
+        title = issue["title"]
+        wip = "WIP: {0}: {1}".format(hostname, title)
```

Comment:
Should we use a different prefix ei (WIP-BOT), since there may be times we want to run tasks on PRs that are WIP